node/src/types/sync_leap.rs

```diff
@@ -290,20 +295,1676 @@ impl FetchItem for SyncLeap {
+            trusted_ancestor_headers: vec![
+                trusted_ancestor_1.header().clone(),
+                trusted_ancestor_2.header().clone(),
+                trusted_ancestor_3.header().clone(),
+            ],
+            signed_block_headers: vec![
+                signed_block_header_with_metadata_1.clone(),
+                signed_block_header_with_metadata_2.clone(),
+                signed_block_header_with_metadata_3.clone(),
+            ],
+        };
+
+        let actual_headers: BTreeSet<_> = sync_leap
+            .switch_blocks_headers()
+            .map(|header| header.block_hash())
+            .collect();
+        let expected_headers: BTreeSet<_> = [
+            trusted_ancestor_1.clone(),
+            signed_block_1.clone(),
+            signed_block_2.clone(),
+        ]
+        .iter()
+        .map(|block| *block.hash())
+        .collect();
+        assert_eq!(expected_headers, actual_headers);
+
+        // Also test when the trusted block is a switch block.
+        let trusted_block = Block::random_switch_block(&mut rng);
+        let sync_leap = SyncLeap {
+            trusted_ancestor_only: false,
+            trusted_block_header: trusted_block.header().clone(),
+            trusted_ancestor_headers: vec![
+                trusted_ancestor_1.header().clone(),
+                trusted_ancestor_2.header().clone(),
+                trusted_ancestor_3.header().clone(),
+            ],
+            signed_block_headers: vec![
+                signed_block_header_with_metadata_1,
+                signed_block_header_with_metadata_2,
+                signed_block_header_with_metadata_3,
+            ],
+        };
+        let actual_headers: BTreeSet<_> = sync_leap
+            .switch_blocks_headers()
+            .map(|header| header.block_hash())
+            .collect();
+        let expected_headers: BTreeSet<_> = [
+            trusted_block,
+            trusted_ancestor_1,
+            signed_block_1,
+            signed_block_2,
+        ]
+        .iter()
+        .map(|block| *block.hash())
+        .collect();
+        assert_eq!(expected_headers, actual_headers);
+    }
+
+    #[test]
+    fn should_return_highest_block_header_from_trusted_block() {
+        // Chain
+        // 0   1   2   3   4   5   6   7   8   9   10   11
+        // S           S           S           S
+        let switch_blocks = [0, 3, 6, 9];
+
+        let mut rng = TestRng::new();
+
+        let query = 5;
+        let trusted_ancestor_headers = [4, 3];
+        let signed_block_headers = [6, 9, 11];
+        let add_proofs = true;
+        let sync_leap = make_test_sync_leap(
+            &mut rng,
+            &switch_blocks,
+            query,
+            &trusted_ancestor_headers,
+            &signed_block_headers,
+            add_proofs,
+        );
+
+        // `sync_leap` is a well formed SyncLeap structure for the test chain. We can use the blocks
```

Comment:
For this comment, I'd maybe mention that while it's well-formed, it's invalid.

Same point for the copies of this comment in the following tests.

Reply:
It refers to the `sync_leap` created above, which is actually valid.